server/src/main.rs

```diff
@@ -1,145 +1,38 @@
+extern crate ws;
+extern crate uuid;
+extern crate time;
 extern crate iron;
 extern crate router;
 extern crate mount;
 extern crate staticfile;
-extern crate uuid;
-extern crate time;
-extern crate ws;
+
+mod web;
+mod sock;
 
 use std::env;
 use std::thread;
-use std::path::Path;
-use std::sync::{Arc,Mutex};
-use std::collections::HashMap;
-
-use uuid::Uuid;
-
-use iron::prelude::*;
-
-use mount::Mount;
-use staticfile::Static;
-
-#[derive(Debug, Clone, Copy)]
-pub struct Connection {
-    uuid: Uuid,
-}
-
-impl std::fmt::Display for Connection {
-    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-        write!(f, "conn{{{}}}", self.uuid)
-    }
-}
 
 fn main() {
-    let connections = Arc::new(Mutex::new(HashMap::new()));
-
-    struct WSServer {
-        out: ws::Sender,
-        connections: Arc<Mutex<HashMap<Uuid, Connection>>>
-    }
-
-    impl ws::Handler for WSServer {
-        fn on_request(&mut self, req: &ws::Request) -> ws::Result<ws::Response> {
-            let mut resp = ws::Response::from_request(req).unwrap();
-
-            let mut conn_map = self.connections.lock().unwrap();
-
-            println!("ws:req[{}]", time::precise_time_ns());
-
-            let new_uuid = Uuid::new_v4();
-            let new_conn = Connection {
-                uuid: new_uuid
-            };
-            conn_map.insert(new_conn.uuid, new_conn);
-            println!("wsconnect: new user with uuid {}", new_uuid);
-            println!("{} connected users", conn_map.len());
-
-            println!("wsconnect: putting cookie: bzwf_anon_wstx={}", new_conn);
-            let new_bzwf_cookie_str = format!("bzwf_anon_wstx={}", new_conn);
-            {
-                let headers = resp.headers_mut();
-                headers.push((
-                    String::from("Set-Cookie"),
-                    new_bzwf_cookie_str.as_bytes().to_vec()
-                ));
-            }
-
-            Ok(resp)
-        }
-
-        fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
-            // TODO
-            println!("ws:rcv[{}]: user with uuid {} sent ws msg {}",
-                     time::precise_time_ns(),
-                     "[[[ we don't have uuids in ws yet ]]]",
-                     msg);
-            self.out.send("I hear you loud and clear").unwrap();
-            Ok(())
+    let domain : String = match env::var("DOMAIN") {
+        Ok(val) => val,
+        Err(_)  => {
+            println!("* DOMAIN not set; using localhost.");
+            String::from("localhost")
         }
+    };
 
-        fn on_open(&mut self, _: ws::Handshake) -> ws::Result<()> {
-            // TODO
-            println!("ws:opn[{}]: user with uuid {} opened ws cxn",
-                     time::precise_time_ns(),
-                     "[[[ we don't have uuids in ws yet ]]]");
-            Ok(())
-        }
+    let tcp_thread_handle = {
+        let domain = domain.clone();
```

Comment:
Turns out it's such a common pattern in Rust to clone something that it's a pain in the ass to go around always renaming it; as such, Rust allows [variable redeclarations (AKA shadowing)](https://doc.rust-lang.org/book/second-edition/ch03-01-variables-and-mutability.html#shadowing) (presumably for this purpose).